utils/test_helpers.go

```diff
@@ -1,27 +1,228 @@
 package utils
 
 import (
+	"bytes"
+	"errors"
 	"fmt"
-	fuzz "github.com/AdaLogics/go-fuzz-headers"
+	"os"
 	"reflect"
-	"testing"
+	"runtime"
+	"runtime/debug"
+	"sync"
+	"time"
+
+	fuzz "github.com/AdaLogics/go-fuzz-headers"
 )
 
+type T struct {
+	parent   *T
+	name     string
+	mu       sync.RWMutex
+	skipped  bool
+	failed   bool
+	finished bool
+	cleanups []func()
+	output   bytes.Buffer
+}
+
+// Most of the T functions are copied from the stdlib
+
+var errNilPanicOrGoexit = errors.New("test executed panic(nil) or runtime.Goexit")
+
+type panicHandling int
+
+const (
+	normalPanic panicHandling = iota
+	recoverAndReturnPanic
+)
+
+func (t *T) log(s string) { _, _ = t.output.WriteString(s) }
+func (t *T) Cleanup(f func()) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.cleanups = append(t.cleanups, f)
+}
+func (t *T) Deadline() (deadline time.Time, ok bool) { return time.Time{}, false }
+func (t *T) Log(args ...any) {
+	t.log(fmt.Sprintln(args...))
+}
+func (t *T) Logf(format string, args ...any) {
+	t.log(fmt.Sprintf(format, args...))
+}
+func (t *T) Error(args ...any) {
+	t.log(fmt.Sprintln(args...))
+	t.Fail()
+}
+func (t *T) Errorf(format string, args ...any) {
+	t.log(fmt.Sprintf(format, args...))
+	t.Fail()
+}
+func (t *T) Fatal(args ...any) {
+	t.log(fmt.Sprintln(args...))
+	t.FailNow()
+}
+func (t *T) Fatalf(format string, args ...any) {
+	t.log(fmt.Sprintf(format, args...))
+	t.FailNow()
+}
+func (t *T) Skip(args ...any) {
+	t.log(fmt.Sprintln(args...))
+	t.SkipNow()
+}
+func (t *T) Skipf(format string, args ...any) {
+	t.log(fmt.Sprintf(format, args...))
+	t.SkipNow()
+}
+func (t *T) Fail() {
+	if t.parent != nil {
+		t.parent.Fail()
+	}
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.failed = true
+}
+func (t *T) FailNow() {
+	t.Fail()
+	t.mu.Lock()
+	t.finished = true
+	t.mu.Unlock()
+	runtime.Goexit()
```

Comment:
Not sure this is the right approach for libFuzzer. I think `os.Exit()` would be better here.

Reply:
This is copy-pasted from `testing`. IMO the behavior of a binary built with go-118-fuzz-build should be as similar to the native fuzzing support as we can achieve with a reasonable effort. `runtime.Goexit` behaves similarly to a panic in that it calls deferred functions. Functions passed to `T.Cleanup()` are called by a deferred closure. If I replaced this with `os.Exit()`, cleanup functions would not be called if a test fails. Fuzz tests probably should not be doing things like creating temp files, but if someone does that I want `T.Cleanup()` to work the same as it does for native fuzzing.

Reply:
Has it been tested in a libFuzzer fuzzer?